src/test/java/org/commcare/test/utilities/CasePurgeTest.java

```diff
@@ -24,58 +28,86 @@
  *
  * Created by ctsims on 10/13/2015.
  */
+@RunWith(Parameterized.class)
 public class CasePurgeTest {
 
-    private final String name;
-
-    private final HashSet<String> cases = new HashSet<>();
-
-    private final HashSet<String> ownedCases = new HashSet<>();
-    private final HashSet<String> closedCases = new HashSet<>();
-    private final HashSet<String> outcome = new HashSet<>();
-
-    private final ArrayList<String[]> indices = new ArrayList<>();
-
-    public static ArrayList<CasePurgeTest> getTests(String resourceName) {
+    @Parameterized.Parameters(name = "{0}")
+    public static Iterable<Object[]> testData() {
         try {
-            ArrayList<CasePurgeTest> runners = new ArrayList<>();
-            JSONArray tests = new JSONArray(TestHelpers.getResourceAsString(resourceName));
-            for (int i = 0; i < tests.length(); ++i) {
-                JSONObject root = tests.getJSONObject(i);
-                runners.add(new CasePurgeTest(root));
+            JSONArray fullTestResource =
+                    new JSONArray(TestHelpers.getResourceAsString("/case_relationship_tests.json"));
+            List<Object[]> listOfParameterSets = new ArrayList<>();
+            for (int i = 0; i < fullTestResource.length(); ++i) {
+                JSONObject root = fullTestResource.getJSONObject(i);
+                listOfParameterSets.add(parseParametersFromJSONObject(root));
             }
-            return runners;
+            return listOfParameterSets;
         } catch (IOException | JSONException e) {
             RuntimeException failure =
-                    new RuntimeException("Failed to parse input for test: " + resourceName);
+                    new RuntimeException("Failed to parse input for CasePurgeTest");
             failure.initCause(e);
             throw failure;
         }
     }
 
-    private CasePurgeTest(JSONObject root) throws JSONException {
-        name = root.getString("name");
-        if (root.has("cases")) {
-            getCases(root.getJSONArray("cases"), cases);
+    private static Object[] parseParametersFromJSONObject(JSONObject root) {
+        Object[] parameters = new Object[7];
+        parameters[0] = root.getString("name");
+
+        String[] jsonArrayKeys =
+                new String[]{"cases", "owned", "closed", "subcases", "extensions", "outcome"};
+        for (int i = 0; i < jsonArrayKeys.length; i++) {
+            addJSONArrayIfPresent(root, i+1, jsonArrayKeys[i], parameters);
+        }
+
+        return parameters;
+    }
+
+    private static void addJSONArrayIfPresent(JSONObject root, int index, String key,
+                                              Object[] parameterSet) {
+        if (root.has(key)) {
+            parameterSet[index] = root.getJSONArray(key);
```

Comment:
Hm, that's a good point. The issues with that are a) what you mentioned about wanting to make sure those objects are JSONArrays, and b) also that "name" is required while the rest are optional, so I'd want to handle it slightly differently anyway. I think given those 2 things, I'm going to leave it the way it is, despite being a bit convoluted.